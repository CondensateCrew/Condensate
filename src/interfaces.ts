// DATA INTERFACES
export interface UserLoginPosting {
  email: string,
  password: string
}

export interface UserLoginReceived {
  readonly id: number,
  readonly firstName: string,
  readonly lastName: string
}

export interface UserSignupPosting {
  email: string,
  password: string,
  firstName: string,
  lastName: string,
  id?: number
}

export interface Action {
  id: number,
  name: string
}

export interface Brainstorm {
  id: number,
  question: string,
  idea: string,
  action: string,
  isGenius: boolean,
  categories: Category[]
}

export interface Category {
  id: number,
  name: string
}

export interface Insight {
  id: number,
  question: string,
  answers: string[]
}

export interface WordContext {
  type: string,
  connectedWord: string[]
}

export interface RandomWordCollection {
  [key: string]: WordContext
}

export interface AppStore {
  user: UserLoginReceived,
  allBrainstorms: Brainstorm[],
  actions: Action[],
  categories: Category[],
  readonly randomWordCollections: RandomWordCollection[],
  readonly questionTemplates: string[],
  readonly secretSauce: string[],
  currentBrainstorm: Brainstorm,
  chosenWords: string[],
  insights: Insight[],
  query: string,
  filter: string,
  timeEnded: boolean,
  instructionsEnded: boolean
}

// ACTIONS OBJECTS

export interface ActionObject {
  type: string,
  [key: string]: any
}

export interface IAddUserAction {
  type: 'ADD_USER',
  user: UserLoginReceived
}

export interface ILogoutUserAction {
  type: 'LOG_OUT_USER'
}

export interface IAddAllActionsAction {
  type: 'ADD_ALL_ACTIONS',
  actions: Action[]
}

export interface IAddNewActionAction {
  type: 'ADD_NEW_ACTION',
  action: Action
}

export interface IAddAllBrainstormsAction {
  type: 'ADD_ALL_BRAINSTORMS',
  brainstorms: Brainstorm[]
}

export interface IChangeBrainstormAction {
  type: 'TOGGLE_BRAINSTORMS',
  id: number
}

export interface IAddNewBrainstormAction {
  type: 'ADD_BRAINSTORM',
  brainstorm: Brainstorm
}

export interface IDeleteBrainstormAction {
  type: 'DELETE_BRAINSTORM',
  id: number
}

export interface IAddAllCategoriesAction {
  type: 'ADD_ALL_CATEGORIES',
  categories: Category[]
}

export interface IAddNewCategoryAction {
  type: 'ADD_CATEGORY',
  category: Category
}

export interface IAddRandomWordCollectionsAction {
  type: 'ADD_ALL_COLLECTIONS',
  collections: RandomWordCollection[]
}

export interface IAddQuestionTemplatesAction {
  type: 'ADD_ALL_TEMPLATES',
  templates: string[]
}

export interface IAddSecretSauceAction {
  type: 'ADD_SECRETE_SAUSE',
  secretSauce: string[]
}

export interface IAddCurrentBrainstormAction {
  type: 'ADD_CURRENT_BRAINSTORM',
  currentBrainstorm: Brainstorm
}

export interface IRemoveCurrentBrainstormAction {
  type: 'REMOVE_CURRENT_BRAINSTORM'
}

export interface IAddChosenWordAction {
  type: 'ADD_WORD',
  chosenWord: string
}

export interface IRemoveChosenWordAction {
  type: 'REMOVE_WORD',
  chosenWord: string
}

export interface IAddInsightAction {
  type: 'ADD_INSIGHT',
  insight: Insight
}

export interface IAddQueryAction {
  type: 'ADD_QUERY',
  query: string
}

export interface IRemoveQueryAction {
  type: 'REMOVE_QUERY'
}

export interface IAddFilterAction {
  type: 'ADD_FILTER',
  filter: string
}

export interface IRemoveFilterAction {
  type: 'REMOVE_FILTER'
}

export interface IEndTimeAction {
  type: 'END_TIME'
}

export interface IReverseTimeAction {
  type: 'REVERSE_TIME'
}

export interface IEndInstructionsAction {
  type: 'END_INSTRUCTIONS'
}

export interface IReverseInstructionsAction {
  type: 'REVERSE_INSTRUCTIONS'
}
