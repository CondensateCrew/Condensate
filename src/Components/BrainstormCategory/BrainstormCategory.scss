@import '../../css/index.scss';
@import '../../css/variables.scss';

.category-btn {
  @extend %buttons;
  margin: $tiny-space;
  &:hover {
    background: $hover-btn-gradient;
    box-shadow: $external-shadow, $regular-focus-shadow;
  }
  &:active {
    box-shadow: $inner-shadow;
  }
}

.active {
  box-shadow:$regular-btn-active-shadow;
  &:hover {
    background: $hover-btn-gradient;
    box-shadow: $inner-shadow;
  }
}
