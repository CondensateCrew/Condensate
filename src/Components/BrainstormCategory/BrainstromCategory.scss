@import '../../css/index.scss';
@import '../../css/variables.scss';

.category-btn {
  @extend %buttons;
  margin: .2rem;
  &:hover {
    background: $hover-btn-gradient;
    box-shadow: $external-shadow, $regular-focus-shadow;
  }
}