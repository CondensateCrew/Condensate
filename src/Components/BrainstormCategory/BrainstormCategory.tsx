import React, { useState, useEffect } from 'react';
import './BrainstormCategory.scss';
import { IBrainstormForm, Category } from '../../interfaces';

interface Props {
  category: Category,
  formState: IBrainstormForm,
  setCategory: (formState: IBrainstormForm) => void
}

const BrainstormCategory:React.FC<Props> = (props) => {
  const { category, formState, setCategory } = props;
  const [ active, setActive ] = useState<boolean>(false);

  const toggleActive = (): void => setActive(!active);

  const addRemoveCategory = (): void => {
    const { categories } = formState;
    const newCategories: Category[] = active
      ? [...categories, category]
      : categories.filter((ctg:Category) => ctg !== category);

    setCategory({...formState, categories: newCategories});
  };

  useEffect(() => {
    if(formState.categories.includes(category)) {
      setActive(true);
    }
  }, [ category, formState ]);

  useEffect(addRemoveCategory, [ active ]);

  useEffect(() => {
    if (formState.reset) return setActive(false);
  }, [formState]);

  const btnClassName: string = active ? 'category-btn active' : 'category-btn';

  return (
    <button type='button' className={btnClassName} value={category.name} onClick={toggleActive}>{category.name}</button>
  );
}

export default BrainstormCategory;
