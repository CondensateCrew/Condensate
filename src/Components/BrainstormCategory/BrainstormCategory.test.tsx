/**
 * @jest-environment jsdom
 */
import React from 'react';
import { mount } from 'enzyme';
import BrainstormCategory from './BrainstormCategory';
import { IBrainstormForm } from 'interfaces';

describe('BrainstormCategory', () => {
  let wrapper: any;

  const mockFormState: IBrainstormForm = {
    question: '',
    categories: [],
    action: {id: 1, action: 'create an app'},
    reset: false
  };

  const mockCategory = {
    id: 1,
    name: 'Technology'
  };

  const mockSetCategory = jest.fn();

  const mockProps = {
    category: mockCategory,
    formState: mockFormState,
    setCategory: mockSetCategory
  }

  beforeEach(() => {
    wrapper = mount(
      <BrainstormCategory {...mockProps}/>
    );
  })

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a different snapshot after a click event', () => {
    wrapper.find('button').simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  it('should call addRemoveCategory when inactive button is clicked', () => {
    const mockArgument = {
      ...mockFormState,
      categories: [...mockFormState.categories, mockCategory]
    };
    wrapper.find('button').simulate('click');

    expect(mockSetCategory).toHaveBeenCalledWith(mockArgument);
  });

  it('should call addRemoveCategory when active button is clicked', () => {
    const mockArgument = {
      ...mockFormState,
      categories: []
    };

    wrapper.find('button').simulate('click');
    wrapper.find('button').simulate('click');

    expect(mockSetCategory).toHaveBeenCalledWith(mockArgument);
  });

  it('should reset local state when formState.reset is true', () => {
    wrapper.find('button').simulate('click');
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({
      category: mockCategory,
      formState: {...mockFormState, reset: true},
      mockSetCategory: jest.fn()
    });

    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
