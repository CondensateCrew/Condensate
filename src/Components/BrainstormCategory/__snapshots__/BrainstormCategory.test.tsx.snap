// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrainstormCategory should match the snapshot 1`] = `
<BrainstormCategory
  category={
    Object {
      "id": 1,
      "name": "Technology",
    }
  }
  formState={
    Object {
      "action": Object {
        "action": "create an app",
        "id": 1,
      },
      "categories": Array [],
      "question": "",
      "reset": false,
    }
  }
  setCategory={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "action": Object {
              "action": "create an app",
              "id": 1,
            },
            "categories": Array [],
            "question": "",
            "reset": false,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <button
    className="category-btn"
    onClick={[Function]}
    type="button"
    value="Technology"
  >
    Technology
  </button>
</BrainstormCategory>
`;

exports[`BrainstormCategory should render a different snapshot after a click event 1`] = `
<BrainstormCategory
  category={
    Object {
      "id": 1,
      "name": "Technology",
    }
  }
  formState={
    Object {
      "action": Object {
        "action": "create an app",
        "id": 1,
      },
      "categories": Array [],
      "question": "",
      "reset": false,
    }
  }
  setCategory={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "action": Object {
              "action": "create an app",
              "id": 1,
            },
            "categories": Array [],
            "question": "",
            "reset": false,
          },
        ],
        Array [
          Object {
            "action": Object {
              "action": "create an app",
              "id": 1,
            },
            "categories": Array [
              Object {
                "id": 1,
                "name": "Technology",
              },
            ],
            "question": "",
            "reset": false,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <button
    className="category-btn active"
    onClick={[Function]}
    type="button"
    value="Technology"
  >
    Technology
  </button>
</BrainstormCategory>
`;

exports[`BrainstormCategory should reset local state when formState.reset is true 1`] = `
<BrainstormCategory
  category={
    Object {
      "id": 1,
      "name": "Technology",
    }
  }
  formState={
    Object {
      "action": Object {
        "action": "create an app",
        "id": 1,
      },
      "categories": Array [],
      "question": "",
      "reset": false,
    }
  }
  setCategory={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "action": Object {
              "action": "create an app",
              "id": 1,
            },
            "categories": Array [],
            "question": "",
            "reset": false,
          },
        ],
        Array [
          Object {
            "action": Object {
              "action": "create an app",
              "id": 1,
            },
            "categories": Array [
              Object {
                "id": 1,
                "name": "Technology",
              },
            ],
            "question": "",
            "reset": false,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <button
    className="category-btn active"
    onClick={[Function]}
    type="button"
    value="Technology"
  >
    Technology
  </button>
</BrainstormCategory>
`;

exports[`BrainstormCategory should reset local state when formState.reset is true 2`] = `
<BrainstormCategory
  category={
    Object {
      "id": 1,
      "name": "Technology",
    }
  }
  formState={
    Object {
      "action": Object {
        "action": "create an app",
        "id": 1,
      },
      "categories": Array [],
      "question": "",
      "reset": true,
    }
  }
  mockSetCategory={[MockFunction]}
  setCategory={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "action": Object {
              "action": "create an app",
              "id": 1,
            },
            "categories": Array [],
            "question": "",
            "reset": false,
          },
        ],
        Array [
          Object {
            "action": Object {
              "action": "create an app",
              "id": 1,
            },
            "categories": Array [
              Object {
                "id": 1,
                "name": "Technology",
              },
            ],
            "question": "",
            "reset": false,
          },
        ],
        Array [
          Object {
            "action": Object {
              "action": "create an app",
              "id": 1,
            },
            "categories": Array [],
            "question": "",
            "reset": true,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <button
    className="category-btn"
    onClick={[Function]}
    type="button"
    value="Technology"
  >
    Technology
  </button>
</BrainstormCategory>
`;
